### ISSUE TITLE
redfish_facts not working for ProLiant DL380 Gen9

### ISSUE THREAD

**Issue by mariolenz** [ansible/ansible#48296]
##### SUMMARY
I am trying to gather BiosAttributes with redfish_facts from a ProLiant DL380 Gen9 but the task failes. I think this is because the module connects to https://ilo_ip/redfish/v1 and iLO redirects to https://ilo_ip/redfish/v1/ with a trailing slash. To me, it looks like redfish_facts doesn's follow the redirect.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
!component = lib/ansible/module_utils/redfish_utils.py

or

!component = lib/ansible/module_utils/urls.py

##### ANSIBLE VERSION
```
ansible 2.7.1
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.6/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.6.5 (default, Apr 21 2018, 13:38:10) [GCC 6.3.0]
```

##### CONFIGURATION
```

```

##### OS / ENVIRONMENT
HP ProLiant DL380 Gen9, iLO Firmware Version 2.61

##### STEPS TO REPRODUCE
Run this Playbook:

```
- hosts: localhost
  connection: local
  tasks:
    - name: Get Bios Attributes
      redfish_facts:
        category: Systems
        command: GetBiosAttributes
        baseuri: "{{ ilo_ip }}"
        user: "{{ user }}"
        password: "{{ password }}"
```

##### EXPECTED RESULTS
Well, the BiosAttributes...

##### ACTUAL RESULTS
In function _find_systems_resource in redfish_utils.py, `self.get_request(self.root_uri + uri)` results in `{'msg': 'HTTP Error: 308', 'ret': False}`:
```
 [WARNING]: Unable to parse /etc/ansible/hosts as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [localhost] *****************************************************************************************

TASK [Gathering Facts] **********************************************************************************
ok: [localhost]

TASK [Get Bios Attributes] ******************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "HTTP Error: 308"}
        to retry, use: --limit @/playbooks/test.retry

PLAY RECAP *******************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1
```

**Comment by ansibot** [comment]
Hi @mariolenz, thank you for submitting this issue!

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/module_utils/aws/urls.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/aws/urls.py)
* [lib/ansible/module_utils/redfish_utils.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/redfish_utils.py)
* [lib/ansible/module_utils/urls.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/urls.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @billdodd @jose-delarosa @mraineri @tomasg2012
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by mariolenz** [comment]
Another machine with Python 2:

```
ansible 2.7.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, May 31 2018, 09:41:32) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```

Same result:

```
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not
match 'all'


PLAY [localhost] *****************************************************************************************

TASK [Gathering Facts] **********************************************************************************
ok: [localhost]

TASK [Get Bios Attributes] ******************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "HTTP Error: 308"}
        to retry, use: --limit @/root/test.retry

PLAY RECAP *******************************************************************************************
```

**Comment by mariolenz** [comment]
!component - lib/ansible/module_utils/aws/urls.py

**Comment by jyundt** [comment]
@mariolenz I think you are hitting #45704. 

Can you try with `devel` and see if that fixes your issue? If so, I submitted a backport request (#47822) which should land in 2.7.2.

**Comment by mariolenz** [comment]
@jyundt Yes, that fixes my issue. Thanks!

duplicate_of #45513

**Comment by jose-delarosa** [comment]
So close?

Thanks @jyundt for the suggestion.

**Comment by mariolenz** [comment]
duplicate_of #45513

**Comment by jyundt** [comment]
Hmm, upon further examination of this issue, it may be worth reopening it. I think the redirect issue is addressed with #45704, however, even with that patch applied I'm still unable to gather redfish_facts from a Gen9 / iLO4 system.

I just discovered that HP Gen9 / iLO4 systems do **not** store BIOS Attrs in the `Attributes` key in  `/redfish/v1/Systems/$YOURSYSTEM/Bios`

This ends up breaking for [get_bios_attributes()](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/redfish_utils.py#L552-L553).

Example of a Gen10 / iLO5 system (BIOS 1.43 / BMC 1.35):

```
$ curl -skL -u root:$BMC_PASSWORD https://$GEN10BMC/redfish/v1/Systems/1/Bios/ | jq '.Attributes'
{
  "iSCSIPolicy": "SoftwareInitiator",
  "XptPrefetcher": "Auto",
  "WorkloadProfile": "GeneralPowerEfficientCompute",
  "WakeOnLan": "Enabled",
  "VlanPriority": 0,
  "VlanId": 0,
  "VlanControl": "Disabled",
  "VirtualSerialPort": "Com2Irq3",
  "VirtualInstallDisk": "Disabled",
<---SNIP--->
  "ProcHyperthreading": "Enabled",
  "ProcTurbo": "Enabled",
  "ProcVirtualization": "Enabled",
  "ProcX2Apic": "Enabled"
}
$ 
```

Example of a Gen9 / iLO4 system (BIOS 2.60 / BMC 2.61)
```
$ curl -sk -L -u root:$BMC_PASSWORD https://$GEN9BMC/redfish/v1/Systems/1/Bios | jq '.Attributes'
null
$ 
```

**Comment by mraineri** [comment]
What is in the payload for `https://$GEN9BMC/redfish/v1/Systems/1/Bios`? If it's listing its attributes in something other than the "Attributes" property, then it's not conformant with the Redfish schema for the Bios resource.

**Comment by jyundt** [comment]
Correct, this is not stored in `Attributes` property and does not appear to be DMTF compliant.

Based on their publish [iLO4 API spec](https://hewlettpackard.github.io/ilo-rest-api-docs/ilo4/#hpbios), this seems to be intentional. I'm reaching out to HPE for additional clarification.

Payload example:
```
$ curl -sk -L -u root:$BMC_PASSWORD https://$GEN9BMC/redfish/v1/Systems/1/Bios | jq '.'
{
  "links": {
    "self": {
      "href": "/rest/v1/systems/1/bios"
    },
    "iScsi": {
      "href": "/rest/v1/systems/1/bios/iScsi"
    },
    "Settings": {
      "href": "/rest/v1/systems/1/bios/Settings"
    },
    "Mappings": {
      "href": "/rest/v1/systems/1/bios/Mappings"
    },
    "Boot": {
      "href": "/rest/v1/systems/1/bios/Boot"
    },
    "BaseConfigs": {
      "href": "/rest/v1/systems/1/bios/BaseConfigs"
    }
  },
  "WakeOnLan": "Enabled",
  "VlanPriority": 0,
  "VlanId": 0,
  "VlanControl": "Disabled",
  "VirtualSerialPort": "Com2Irq3",
  "VirtualInstallDisk": "Disabled",
  "UtilityLang": "English",
  "UsbControl": "UsbEnabled",
  "UsbBoot": "Enabled",
  "Usb3Mode": "Auto",
  "UrlBootFile": "",
  "UefiShellStartupUrl": "",
  "UefiShellStartupLocation": "Auto",
  "UefiShellStartup": "Disabled",
  "UefiShellBootOrder": "Disabled",
  "UefiPxeBoot": "Auto",
  "UefiOptimizedBoot": "Disabled",
  "Type": "HpBios.1.2.0",
  "TpmType": "NoTpm",
  "TpmState": "NotPresent",
  "TimeZone": "Utc0",
  "TimeFormat": "Utc",
  "ThermalShutdown": "Enabled",
  "ThermalConfig": "OptimalCooling",
  "Ipv4SubnetMask": "0.0.0.0",
  "Ipv4SecondaryDNS": "0.0.0.0",
  "Ipv4PrimaryDNS": "0.0.0.0",
  "Ipv4Gateway": "0.0.0.0",
  "Ipv4Address": "0.0.0.0",
  "IoNonPostedPrefetching": "Enabled",
  "InternalSDCardSlot": "Enabled",
  "IntelligentProvisioning": "Enabled",
  "IntelQpiPowerManagement": "Enabled",
  "IntelQpiLinkEn": "Auto",
  "IntelQpiFreq": "Auto",
  "IntelProcVtd": "Enabled",
  "IntelPerfMonitoring": "Disabled",
  "IntelNicDmaChannels": "Enabled",
  "IntelDmiLinkFreq": "Auto",
  "HwPrefetcher": "Enabled",
  "FanInstallReq": "EnableMessaging",
  "FanFailPolicy": "Shutdown",
  "FCScanPolicy": "AllTargets",
  "F11BootMenu": "Enabled",
  "ExtendedMemTest": "Disabled",
  "ExtendedAmbientTemp": "Disabled",
  "EraseUserDefaults": "No",
  "EnergyPerfBias": "BalancedPerf",
  "EmsConsole": "Disabled",
  "EmbeddedUserPartition": "Disabled",
  "EmbeddedUefiShell": "Enabled",
  "EmbeddedSerialPort": "Com1Irq4",
  "EmbeddedSata": "Ahci",
  "EmbeddedDiagsMode": "Auto",
  "EmbeddedDiagnostics": "Enabled",
  "EmbVideoConnection": "Auto",
  "BootOrderPolicy": "RetryIndefinitely",
  "BootMode": "LegacyBios",
  "AutoPowerOn": "RestoreLastState",
  "AttributeRegistry": "HpBiosAttributeRegistryP89.1.1.60",
  "AssetTagProtection": "Unlocked",
  "AsrTimeoutMinutes": "10",
  "AsrStatus": "Enabled",
  "AdvancedMemProtection": "AdvancedEcc",
  "AcpiRootBridgePxm": "Enabled",
  "AcpiSlit": "Enabled",
  "AdjSecPrefetch": "Enabled",
  "AdminEmail": "",
  "AdminName": "",
  "AdminOtherInfo": "",
  "AdminPassword": null,
  "AdminPhone": "",
  "ChannelInterleaving": "Enabled",
  "CollabPowerControl": "Enabled",
  "ConsistentDevNaming": "LomsOnly",
  "CustomPostMessage": "",
  "DaylightSavingsTime": "Disabled",
  "DcuIpPrefetcher": "Enabled",
  "DcuStreamPrefetcher": "Enabled",
  "Description": "This is the Platform/BIOS Configuration (RBSU) Current Settings",
  "Dhcpv4": "Enabled",
  "DynamicPowerCapping": "Auto",
  "DynamicPowerResponse": "Fast",
  "EmbNicEnable": "Enabled",
  "EmbSas1Boot": "AllTargets",
  "EmbSasEnable": "Enabled",
  "EmbSata1Enable": "Enabled",
  "EmbSata2Enable": "Enabled",
  "Ipv6Duid": "Auto",
  "MaxMemBusFreqMHz": "Auto",
  "MaxPcieSpeed": "MaxSupported",
  "MemFastTraining": "Enabled",
  "MinProcIdlePkgState": "C6Retention",
  "MinProcIdlePower": "C6",
  "MixedPowerSupplyReporting": "Enabled",
  "Modified": "2018-11-15T16:37:57+00:00",
  "Name": "BIOS Current Settings",
  "NetworkBootRetry": "Enabled",
  "NicBoot1": "NetworkBoot",
  "NicBoot2": "Disabled",
  "NicBoot3": "Disabled",
  "NicBoot4": "Disabled",
  "NmiDebugButton": "Enabled",
  "NodeInterleaving": "Disabled",
  "NumaGroupSizeOpt": "Clustered",
  "OldAdminPassword": null,
  "OldPowerOnPassword": null,
  "PciBusPadding": "Enabled",
  "PciSlot1Enable": "Enabled",
  "PcieExpressEcrcSupport": "Disabled",
  "PostF1Prompt": "Delayed20Sec",
  "PowerButton": "Enabled",
  "PowerOnDelay": "None",
  "PowerOnLogo": "Enabled",
  "PowerOnPassword": null,
  "PowerProfile": "BalancedPowerPerf",
  "PowerRegulator": "DynamicPowerSavings",
  "PreBootNetwork": "Auto",
  "ProcAes": "Enabled",
  "ProcCoreDisable": 0,
  "ProcHyperthreading": "Enabled",
  "ProcNoExecute": "Enabled",
  "ProcTurbo": "Enabled",
  "ProcVirtualization": "Enabled",
  "ProcX2Apic": "Enabled",
  "ProductId": "N4M29A",
  "QpiBandwidthOpt": "Balanced",
  "QpiSnoopConfig": "Standard",
  "RedundantPowerSupply": "BalancedMode",
  "RemovableFlashBootSeq": "ExternalKeysFirst",
  "RestoreDefaults": "No",
  "RestoreManufacturingDefaults": "No",
  "RomSelection": "CurrentRom",
  "SataSecureErase": "Disabled",
  "SaveUserDefaults": "No",
  "SecureBootStatus": "Disabled",
  "SerialConsoleBaudRate": "115200",
  "SerialConsoleEmulation": "Vt100Plus",
  "SerialConsolePort": "Physical",
  "SerialNumber": "##########",
  "ServerAssetTag": "",
  "ServerName": "",
  "ServerOtherInfo": "",
  "ServerPrimaryOs": "",
  "ServiceEmail": "",
  "ServiceName": "",
  "ServiceOtherInfo": "",
  "ServicePhone": "",
  "SettingsResult": {
    "Time": "2018-01-19T20:39:32+00:00",
    "Messages": [
      {
        "MessageID": "Base.1.0:Success",
        "MessageArgs": []
      }
    ],
    "ETag": "A3A6-01-01T00:01:33+00:00"
  },
  "Slot1NicBoot1": "NetworkBoot",
  "Slot1NicBoot2": "Disabled",
  "Sriov": "Enabled"
}
$
```

**Comment by jyundt** [comment]
@mraineri @jose-delarosa 

I  spoke with someone from HPE and this is working as designed, the Gen9 was released prior to the standardization of the BIOS attrs in the redfish spec. Bios attrs are available as an HPE OEM extension, they aren't advertised via "@odata.type=Bios...."

Based on that, any thoughts/suggestions on how to proceed?

Re-open this issue? Open a new issue?

**Comment by jose-delarosa** [comment]
@jyundt, @billdodd developed a POC to integrate/handle OEM extensions. See https://github.com/ansible/ansible/issues/48364. This could be a potential way to handle your problem. I suppose you could open another bug for you to track it, and then provide your own PR once we determine the best way to handle OEM extensions. Just a thought.

**Comment by jyundt** [comment]
After looking at #48364 / #48740, I'll wait for the OEM extension bits. I'll open a separate issue to track this.

In the meantime, I think this Gen9 specific issue can be worked around using the `uri` module.

**Comment by mariolenz** [comment]
@jyundt 
> Hmm, upon further examination of this issue, it may be worth reopening it. I think the redirect issue
> is addressed with #45704, however, even with that patch applied I'm still unable to gather
> redfish_facts from a Gen9 / iLO4 system.

Yes, I've ran into the same problem. However, I closed this issue because I thought the main ansible problem was not following redirects. HPE doesn't follow the standard here but I considered this to be a different issue. (Actually, I tend to the opinion that this is a HPE and not an ansible problem.) I thought that working with OEM specific extensions would be considered out of scope...
